x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/PushDownAndCombineFilters.java

```diff
@@ -143,28 +166,38 @@ private static LogicalPlan pushDownPastJoin(Filter filter, Join join, FoldContex
             LogicalPlan left = join.left();
             LogicalPlan right = join.right();
 
-            // split the filter condition in 3 parts:
-            // 1. filter scoped to the left
-            // 2. filter scoped to the right
-            // 3. filter that requires both sides to be evaluated
-            ScopedFilter scoped = scopeFilter(Predicates.splitAnd(filter.condition()), left, right);
+            // Split the filter condition in 3 parts.
+            // For InlineJoin we use a scoping that allows pushing down filters either to right side only or to both sides.
+            // For the rest of the joins we use the standard scoping:
+            // - filters scoped to the left
+            // - filters scoped to the right
+            // - filter that requires both sides to be evaluated
+            ScopedFilter scoped = join instanceof InlineJoin ij
+                ? scopeInlineStatsFilter(Predicates.splitAnd(filter.condition()), ij)
+                : scopeFilter(Predicates.splitAnd(filter.condition()), left, right);
+
+            // For InlineJoin, in EsqlSession the left-hand side becomes the source for the right-hand side replacing the StubRelation
+            // and, as such, the filters used on the left-hand side are also used on the right-hand side.
+            var pushableToLeftSide = join instanceof InlineJoin ? scoped.commonFilters() : scoped.leftFilters();
+            var pushableToRightSide = scoped.rightFilters();
+            var commonFilters = join instanceof InlineJoin ? scoped.leftFilters() : scoped.commonFilters();
```

Comment:
[Related to the next comment below]
```suggestion
            scoped = join instanceof InlineJoin ? new ScopedFilter(scoped.leftFilters, scoped.commonFilters, scoped.rightFilters) : scoped;
```
..and then keep the rest of the code below unchanged.

Alternatively, do this shuffling in `scopeInlineStatsFilter()` already.

Reply:
I guess the naming here is a bit tricky since, as done now, the filters pushable on the LHS are considered to be the ones taken from the RHS join keys. However, these are the same as the LHS join keys (for InlineJoin).
Basically, we want to push on the LHS of the (inline) join the filters defined on the groupings (which are the join keys). Given that, I think I'd update `scopeInlineStatsFilter()` code to start with the `bothSides` var be initialised with the `filters` param (and not `leftFilters` as is now) and then extract to the `leftFilters` those predicates/filters that are part of the join keys.

This would then allow initialising resulting ScopedFilter instance inline with the usage at this code location where this comment is left.

Reply:
Refactored